Update: legend to provide name to optgroup

scottaohara opened the pull request:
In regards to the updated content model for the select element and its allowed children, an `optgroup` can have a `legend` element as its first child, and this `legend` needs to be able to name the `optgroup` similarly to how a `legend` names a `fieldset`.

see:
https://github.com/whatwg/html/pull/10586

<!--- IF EDITORIAL or CHORE, delete the rest of this template -->

# Test, Documentation and Implementation tracking
Once this PR has been reviewed and has consensus from the working group, tests should be written and issues should be opened on browsers. Add N/A and check when not applicable.

* [ ] "author MUST" tests:
* [ ] "user agent MUST" tests:
* [ ] Browser implementations (link to issue or commit):
   * WebKit:
   * Gecko:
   * Blink:
* [ ] Does this need AT implementations?
* [ ] Related APG Issue/PR:
* [ ] MDN Issue/PR:

keithamus reviewed (approved):
Couple of minor formatting comments, but this LGTM from a spec position, modulo new resolutions from the WHATWG.

keithamus commented on html-aam/index.html:
```diff
@@ -16147,6 +16152,26 @@ <h4>`fieldset` Element Accessible Name Computation</h4>
+            </li>
+            <li>
+              If the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is still empty, then: if the `optgroup` element has a
+              <a data-cite="html/form-elements.html#attr-optgroup-label">label</a> attribute, then use that attribute.
```
Yeah I think it's fine (arguably even good) to keep it linked.

scottaohara replied:
ok. we'll leave it linked for now then and to another pass at cleanup/consistency for other items that maybe should be linked, but aren't right now.

scottaohara commented on html-aam/index.html:
```diff
@@ -16147,6 +16152,26 @@ <h4>`fieldset` Element Accessible Name Computation</h4>
+          <ol>
+            <li>
+              If the `optgroup` element has an <a data-cite="wai-aria-1.2/#aria-label">`aria-label`</a> or an <a data-cite="wai-aria-1.2/#aria-labelledby">`aria-labelledby`</a> attribute the
+              <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is to be calculated using the algorithm defined in
```
label should come before legend
if label is used with legend, then legend should become description for optgroup

scottaohara commented:
@keithamus would you mind taking another look?  I've updated the ordering for the optgroup naming, and added new content to how a description should be calculated.

cc @josepharhar as updates here related to [my comments](https://github.com/whatwg/html/pull/10586#issuecomment-2432523372) in the update content model for customized select PR

scottaohara commented:
examples for tests for how name/desc steps should work
```

<!-- name of optgroup: foo -->
<select>
    
    <optgroup aria-label=foo label=bar>
        <option>test
    </optgroup>
</select>

<!-- name of optgroup: foo 
        desc of optgroup: bar
-->
<select>
    
    <optgroup aria-label=foo title=bar>
        <option>test
    </optgroup>
</select>


<!-- name of optgroup: foo 
        desc of optgroup: bar
-->
<select>
    
    <optgroup label=foo title=bar>
        <option>test
    </optgroup>
</select>

<!-- name of optgroup: bar -->
<select>
    
    <optgroup title=bar>
        <option>test
    </optgroup>
</select>

<!-- name of optgroup: bar -->
<select>
    
    <optgroup label=bar>
        <option>test
    </optgroup>
</select>

<!-- name of optgroup: foo 
        desc of optgroup: bar
-->
<select>
    
    <optgroup aria-labelledby=x label=bar>
        <legend id=x>foo</legend>
        <option>test
    </optgroup>
</select>


<!-- name of optgroup: bar 
        desc of optgroup: foo
-->
<select>
    
    <optgroup label=bar>
        <legend>foo</legend>
        <option>test
    </optgroup>
</select>


<!-- name of optgroup: foo 
     desc of optgroup: bar baz
-->
<select>
    
    <optgroup aria-label=foo label=bar>
        <legend>baz</legend>
        <option>test
    </optgroup>
</select>


<!-- name of optgroup: foo 
     desc of optgroup: bar baz
-->
<select>
    
    <optgroup aria-label=foo label=bar title=ignored>
        <legend>baz</legend>
        <option>test
    </optgroup>
</select>
```

keithamus reviewed (approved):
Still LGTM

scottaohara commented:
put back into the draft state as I need to revise the bits about how a `label` attribute and `legend` element should no longer be expected to render at the same time.

josepharhar commented:
> put back into the draft state as I need to revise the bits about how a `label` attribute and `legend` element should no longer be expected to render at the same time. re: https://issues.chromium.org/issues/378601807

If you have a recommendation for what to do in this case, let me know and/or comment on the crbug. I haven't implemented anything yet to account for the case where both are present.

scottaohara commented:
@josepharhar i think it's totally fine to only have one render / and have a preference for the legend element in this case since it'll allow for authors to do more.

- if only label attribute is specified, render that to be consistent with prior select functionality
- if only legend is specified, render that and if the select is loaded in a browser that doesn't support customization, there is no fallback
- if both are specified, 
  - in browsers that support customization render legend element as the group label and do not render label attribute text
  - in browsers that don't support customization render the fallback label attribute text as the group label

spectranaut commented:
@scottaohara does this need implementation changes? If so I'll add the PR tracking list.

spectranaut commented:
Discussed briefly in last weeks meeting: https://www.w3.org/2025/06/26-aria-minutes.html#58a8

jcsteh commented:
> Gecko: Keith said this is implemented.

I'm not sure this is true. I don't see anything in the Gecko accessibility code that would currently support legend as a labelling child of optgroup. But maybe I'm missing something. @keithamus, can you clarify your thinking here?

jcsteh commented on html-aam/index.html:
```diff
@@ -4697,7 +4697,10 @@ <h4 id="el-legend">`legend`</h4>
-                <div class="relations"><span class="type">Relations:</span> `IA2_RELATION_LABEL_FOR` with the parent <a href="#el-fieldset">`fieldset`</a></div>
+                <div class="relations">
+                  <span class="type">Relations:</span> `IA2_RELATION_LABEL_FOR` with the parent <a href="#el-fieldset">`fieldset`</a> 
+                  or <a href="#el-optgroup">`optgroup`</a>
```
For clarity, I think we need to specify the reverse LABELLED_BY relation somewhere. For example, [we do this for fieldset](https://w3c.github.io/html-aam/#:~:text=Relations%3A%20IA2%5FRELATION%5FLABELLED%5FBY%20with%20the%20first%20instance%20of%20a%20rendered%20child%20legend%20element). We're somewhat inconsistent about where we document the relation. For IA2, we document LABELLED_BY in the `<fieldset>` section, but for UIA, we document LabeledBy in the `<legend>` section. That's an editorial issue we can sort out separately, but I think it does need to be documented somewhere, and I'd suggest documenting the IA2 relation in the `<optgroup>` section, since that's what we do for `<fieldset>`.

jcsteh commented on html-aam/index.html:
```diff
@@ -16400,6 +16427,19 @@ <h3>Accessible Description Computation</h3>
+                    </li>
+                    <li>
+                      or use the <a data-cite="accname-1.2/#mapping_additional_nd_te">text equivalent computation</a> of the subtree of the 
+                      first `legend` element if it was not used as the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a>.
```
We should expose DESCRIBED_BY and DESCRIPTION_FOR relations if we do this. I think we have the same issue for table captions when used as description, but I guess we should address that separately.

jcsteh reviewed (changes requested):
Overall, this looks good, but I think it needs some clarifications.

scottaohara commented:
thanks for the review on this @jcsteh - i'll work on these suggestions asap
